sycl/include/sycl/device.hpp

```diff
@@ -52,6 +55,21 @@ enum class peer_access {
 
 } // namespace ext::oneapi
 
+// Device properties that returns std::string
+// These need special handling to deal with ABI incompatibility
+enum class DeviceProperty {
```

Comment:
I think something like https://github.com/intel/llvm/commit/265eac3d671542e18873b68ac1e1b43c85722a59 should be used instead, but I'm open to discussion with @steffenlarsen / @sergey-semenov .

Reply:
Thanks @aelovikov-intel.
@steffenlarsen @sergey-semenov 
We are discussing the issue of how to call get_info<T>() across two different ABIs.
(Note that this PR is addressing the ABI incompatibility issue between header and libsycl sides)
My approach in this PR is passing T as an an enum value to the other ABI side and recostruct <T>.
Andrei's suggestion is using template<T>, but it requires some changes deep in the libsycl cpp files.
My approach does not need to change other libsycl cpp files.
I think both approaches work. 
Any advice would be appreciated.

Reply:
Byoungro and I talked about this yesterday, but just to summarize here: I strongly lean towards Andrei's approach here, there shouldn't be any need to introduce this enum value. Although one thing I would add is that I think we should keep `detail::string` as "local" as possible, meaning only used to cross the header/library threshold, with the types specified in the .def file and used by strictly internal functions left unchanged.

Reply:
>  I think we should keep detail::string as "local" as possible, meaning only used to cross the header-library threshold, with the types specified in the .def file and used by strictly internal functions left unchanged.

I 100% agree with this and that's why my commit linked above is "[POC]" as it didn't do that. I'm confident we can make use the same template metaprogramming tricks to lift that boundary up in comparison with what I provided there.

Reply:
I finally managed to implement the template approach without enums.
Thank you very much, @sergey-semenov @aelovikov-intel